### ISSUE TITLE
Data Leakage Issue in Football Prediction Model

### ISSUE THREAD

**Issue by f4ieh** [stevenmcsorley/football_prediction#5]
Hi Steven,

I hope this message finds you well. I've been studying your football prediction model on GitHub (stevenmcsorley/football_prediction) and found it quite interesting. However, I've identified some data leakage issues that may be inflating the reported 73.03% accuracy.

Data Leakage Issues Identified:

Rolling Features with Look-ahead: The rolling averages include the current match in the calculation:

df['home_team_strength'] = df.groupby('home_team')['home_goals'].transform(
    lambda x: x.rolling(window, min_periods=1).mean()
)

This gives the model access to information from the match it's trying to predict.
Head-to-Head Statistics: The H2H calculations include all matches (past + future):

h2h_stats = df.groupby(['home_team', 'away_team']).apply(get_h2h_stats)

This incorporates future match results into historical statistics.
Feature Engineering on Complete Dataset: Features are calculated on the entire dataset before the train/test split, allowing future information to leak into training data.
Insufficient Temporal Separation: The 80/20 split without a temporal gap can cause contamination between training and test sets.
Impact Assessment:

I tested your methodology with proper temporal validation and found:

With leakage: ~70% accuracy
Without leakage: ~52% accuracy
Leakage boost: +17-18%
This suggests the true performance of the model is closer to 55-56% rather than 73%.

Recommendations:

Calculate rolling features using only matches prior to the prediction date
Implement proper temporal splits with a gap between train/test
Ensure H2H statistics only use historical data
Perform feature engineering separately on train/test sets
I'm not trying to diminish your work - the model architecture and feature selection are solid. These are common pitfalls in time-series ML that can be easily fixed. Would you be interested in discussing this further or collaborating on a corrected version?

Best regards,

**Comment by zhaokang19940918** [comment]
你修正好了嘛？

**Comment by zhaokang19940918** [comment]
> 嗨，史蒂文，
> 
> 希望您一切安好。我一直在 GitHub (stevenmcsorley/football_prediction) 上研究您的足球预测模型，发现它非常有趣。然而，我发现了一些数据泄露问题，可能导致报告的 73.03% 准确率被夸大。
> 
> 已发现的数据泄漏问题：
> 
> 具有前瞻性的滚动特征：滚动平均值包括计算中的当前匹配：
> 
> df['home_team_strength'] = df.groupby('home_team')['home_goals'].transform( lambda x: x.rolling(window, min_periods=1).mean() )
> 
> 这使得模型能够访问它试图预测的比赛信息。 正面交锋统计数据：正面交锋计算涵盖所有比赛（过去 + 未来）：
> 
> h2h_stats = df.groupby(['主队', '客队']).apply(get_h2h_stats)
> 
> 这将未来的比赛结果纳入历史统计数据。 基于完整数据集的特征工程：在训练/测试分割之前，对整个数据集进行特征计算，这会导致未来信息泄露到训练数据中。 时间分离不足：缺乏时间间隙的 80/20 分割可能会导致训练集和测试集之间出现污染。 影响评估：
> 
> 我通过适当的时间验证测试了您的方法并发现：
> 
> 有泄漏：准确率~70% 无泄漏：准确率~52% 泄漏提升：+17-18% 这表明该模型的真实性能更接近 55-56%，而不是 73%。
> 
> 建议：
> 
> 仅使用预测日期之前的匹配来计算滚动特征； 实施适当的时间分割，并在训练/测试之间留出间隔； 确保 H2H 统计数据仅使用历史数据； 分别在训练集/测试集上执行特征工程 ；我并非想贬低你的工作——你的模型架构和特征选择非常可靠。这些都是时间序列机器学习中常见的陷阱，但可以轻松修复。你有兴趣进一步讨论这个问题，或者合作开发一个修正版本吗？
> 
> 此致，

请问，修正好了嘛？

**Comment by f4ieh** [comment]
enfin oui j'ai fait ma propre version et j'arrive maintenat a m'approcher des 68% mais ya encore du travail

**Comment by zhaokang19940918** [comment]
> enfin oui j'ai fait ma propre version et j'arrive maintenat a m'approcher des 68% mais ya encore du travail

可以，给我分享一下嘛？

**Comment by stevenmcsorley** [comment]
Hi @f4ieh, thanks for the detailed analysis and feedback on the data leakage issues. I appreciate you taking the time to identify these problems. Unfortunately, I'm currently too busy to collaborate on fixes or discuss this further at the moment. However, if you'd like to create a pull request with the suggested improvements, I'd be happy to review it. Thanks again!

**Comment by zhaokang19940918** [comment]
> enfin oui j'ai fait ma propre version et j'arrive maintenat a m'approcher des 68% mais ya encore du travail

Hi, could you share it with me?
